blob/fileblob/fileblob.go

```diff
@@ -606,7 +617,104 @@ func (b *bucket) Delete(ctx context.Context, key string) error {
 	return nil
 }
 
+// SignedURL takes a context, an object key, and a SignedURLOptions struct
+// and returns a string containing a URI and an error
 func (b *bucket) SignedURL(ctx context.Context, key string, opts *driver.SignedURLOptions) (string, error) {
-	// TODO(Issue #546): Implemented SignedURL for fileblob.
-	return "", errNotImplemented
+	if b.opts.URLSigner == nil {
+		return "", errNotImplemented
+	}
+	return b.opts.URLSigner.URLFromKey(ctx, key, opts)
+}
+
+// URLSigner defines an interface for
+// creating and verifying a signed URL for objects
+// in a fileblob bucket. Signed URLs are typically used for
+// granting access to an otherwise-protected resource without
+// requiring further authentication, and callers should take care
+// to restrict the creation of signed URLs as is appropriate
+// for their application.
+type URLSigner interface {
+	// URLFromKey defines how the bucket's object key will be turned
+	// into a signed URL.
+	// URLFromKey takes in a context, object key and options.
+	// It returns a signed URL string, and an error.
+	URLFromKey(ctx context.Context, key string, opts *driver.SignedURLOptions) (string, error)
+
+	// KeyFromURL takes in a context and a URL.
+	// It returns an object key string and a bool.
+	// e.g. (key, true) if the signature is valid (authentic && unexpired),
+	// or (nil, false) if it is invalid.
+	// KeyFromURL must be able to validate a URL from the URLFromKey function.
+	KeyFromURL(ctx context.Context, surl *url.URL) (string, bool)
+}
+
+// URLSignerHMAC uses the crypto/hmac package to create a message authentication code
+// using a sha256 hash
+type URLSignerHMAC struct {
+	urlScheme string
+	urlHost   string
+	urlPath   string
+	secretKey []byte
+}
+
+func NewURLSignerHMAC(urlScheme, urlHost, urlPath, secretKey string) *URLSignerHMAC {
+	return &URLSignerHMAC{
+		urlScheme: urlScheme,
+		urlHost:   urlHost,
+		urlPath:   urlPath,
+		secretKey: []byte(secretKey),
+	}
+}
+
+// URLFromKey uses the scheme and host in URLSignerHMAC, and uses the passed key as the path.
+func (h *URLSignerHMAC) URLFromKey(ctx context.Context, key string, opts *driver.SignedURLOptions) (string, error) {
+	sURL := &url.URL{
+		Host:   h.urlHost,
+		Scheme: h.urlScheme,
+		Path:   h.urlPath,
+	}
+	q := sURL.Query()
+	q.Set("obj", key)
+	expTime := time.Now().Add(opts.Expiry).Unix()
+	q.Set("expiry", strconv.FormatInt(expTime, 10))
+	q.Set("signature", "")
```

Comment:
Nit -- is this needed? If so, it warrants a comment why.

Reply:
I can either set the signature to an empty string here, or unset the "signature" param entirely when the MAC is being checked, since I need to remove the signature value before re-hashing it.